### PR TITLE
Correctly synchronize RecoveryState#Translog internal state

### LINKED ISSUE #80837: [CI] FullClusterRestartIT testRecovery failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/lqmxb2fffkd42/tests/:qa:full-cluster-restart:v8.1.0%23upgradedClusterTest/org.elasticsearch.upgrades.FullClusterRestartIT/testRecovery

**Reproduction line:**
`gradlew ':qa:full-cluster-restart:v8.1.0#upgradedClusterTest' -Dtests.class="org.elasticsearch.upgrades.FullClusterRestartIT" -Dtests.method="testRecovery" -Dtests.seed=37572CF50121804A -Dtests.bwc=true -Dtests.locale=sr-ME -Dtests.timezone=Asia/Barnaul -Druntime.java=17`

**Applicable branches:**
master

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.upgrades.FullClusterRestartIT&tests.test=testRecovery

**Failure excerpt:**
```
java.lang.AssertionError: mismatch while checking for translog recovery
testrecovery 0 existing_store done 0
testrecovery 0 peer           done 0
 expected:<true> but was:<false>

  at __randomizedtesting.SeedInfo.seed([37572CF50121804A:F6A755592C714AED]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:118)
  at org.elasticsearch.upgrades.FullClusterRestartIT.testRecovery(FullClusterRestartIT.java:880)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

### PR DESCRIPTION
**Pull request by arteam**
We must synchronize all access to the internal state, otherwise
there's no read/write memory barriers when deserializing/serialize
the state from the wire.

Fixes #80837

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 1]
```diff
@@ -478,19 +478,23 @@ public Translog() {}
...
-            recovered = in.readVInt();
-            total = in.readVInt();
-            totalOnStart = in.readVInt();
-            totalLocal = in.readVInt();
+            synchronized (this) {
+                recovered = in.readVInt();
+                total = in.readVInt();
+                totalOnStart = in.readVInt();
+                totalLocal = in.readVInt();
+            }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:486]: Looks like this is a constructor. Please correct me if I am wrong, but I do not think it is possible to modify object from other thread while it is still constructed.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:486]: ++ unless the constructor were to spawn a thread this is impossible and there's no reason to at sync here

Comment (arteam) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:486]: That's a good point! Let me remove it.

I guess it'd nice to have some jcstress tests here in order to check that the synchronization is correct.

Comment (arteam) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:486]: Actually, it looks like we still need to have a synchornized block here in order to guarantee visibility. The `recovered`, `total`, `totalOnStart`, `totalLocal` fields are not finals, so we don't have a write fence to guarantee that they will be correctly published for a different thread.

Comment (arteam) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:485]: Background info for that: https://shipilev.net/blog/2014/safe-public-construction/#_safe_initialization

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java (hunk 1 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:485]: :+1: Then I would only suggest to put a comment in constructor explaining why synchronized is needed there

Comment (arteam) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:485]: The issue is with the visibility of the constructor. The compiler doesn't  a write barrier, hence there no is happens-before relationship between writing to to total* fields and reading from them.

Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:485]: It seems like as long as the object is safely published, subsequent synchronized reads should get full visibility of the object fields (see https://stackoverflow.com/a/39393985) But maybe I'm missing something here.

Comment (arteam) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:485]: But that's the crux of the issue!
```
        private int recovered;
        private int total = UNKNOWN;
        private int totalOnStart = UNKNOWN;
        private int totalLocal = UNKNOWN;
```
The fields are neither volatile nor final, so the object is unsafely published.

**Review by original-brownbear — COMMENTED** [review]
I'm having a hard time believing this explains the test failure on x86 tbh. Even though we technically seem to go beyond the guarantees of the JLS here practically. The pattern we use here is widely used across our codebase (construct mutable object then pass to other thread) and has not been the source of issues previously in practice.
This might be suited for raising during a sync :) Maybe we should make an effort to make all our transport object immutable ...

**Comment by arteam** [comment]
@elasticmachine update branch

**Pull request merged**